c++ training/PointersTraining/PointerExercises/S2E2/Source.cpp

```diff
@@ -0,0 +1,43 @@
+#include <iostream>
+
+/* Exercise 2 - Understanding Stack vs Heap
+Let�s compare the memory addresses of the stack and the heap.
+First, create three int variables called stackInt1, stackInt2, and stackInt3. These variables should
+be allocated in the stack. Initialize them to whatever value you want and then print their value
+and memory addresses.
+Then, create three more int variables and call them healpInt1, heapInt2, and heapInt3. These
+variables should be allocated in the heap. Initialize them to whatever value you want and print
+their value and memory addresses.
+What can you notice from the memory addresses in the stack vs the memory addresses in the
+heap? Print the answer to that question to the console (Yep, we�re doing quizzes in exercises
+now).
+Remember to free any memory you allocate in the heap.*/
+
+int main()
+{
+	int StackInt1 = 15;
+	int StackInt2 = 20;
+	int StackInt3 = 16;
+
+	std::cout << "I created three stack variables:\nTheir values are :" << std::endl;
+	std::cout << StackInt1 << std::endl << StackInt2 << std::endl << StackInt3 << std::endl;
+	std::cout << "Their memory addresses are " << std::endl;
+	std::cout << &StackInt1 << std::endl << &StackInt2 << std::endl << &StackInt3 << std::endl << std::endl;
+
+	int* HeapInt1 = new int(36);
+	int* HeapInt2 = new int(5);
+	int* HeapInt3 = new int(11);
+
+	std::cout << "I created three heap variables:\nTheir values are :" << std::endl;
+	std::cout << *HeapInt1 << std::endl << *HeapInt2 << std::endl << *HeapInt3 << std::endl;
+	std::cout << "Their memory addresses are " << std::endl;
+	std::cout << HeapInt1 << std::endl << HeapInt2 << std::endl << HeapInt3 << std::endl << std::endl;
+
+	delete HeapInt1, HeapInt2, HeapInt3;
+
+	std::cout << "The difference between these memory addresses is: essentially, that the variables that areinitialised and allocated in the stack are " << std::endl;
+	std::cout << "being stored continuously one after another, every 4 bytes, which is the size of every integer." << std::endl;
+	std::cout << "Heap allocated integers are, on the other hand, given memory locations that seem much more random, there is in fact not necessarilly " << std::endl;
+	std::cout << "always much correlation between the variables addresses altogether. They might be near each other, however there will not always be" << std::endl;
+	std::cout << "the same distance between one another." << std::endl;
```

Comment:
I can add that the difference is also that the first ones are stored on the stack and the others on the heap, which translates into a difference in data reading speed, in addition to being more than random (because pro yes both have random addresses in memory) is that in one the bytes are separated and it is known that the values will be or will be and in the other it is a promise, that you fulfill it or not is not a matter of dynamic memory

Reply:
I feel like the fact that some are stored or allocated in heap memory and stack memory was many times addressed in the answer.

Reply:
I'm not quite sure of what you're saying there, I haven't found whatsoever yet any information on the speed of data reading would change between these two. There is a difference, however, when you're actually asking to find free memory there is in fact a difference in speed, but only to access free memory, once you have it it's a matter of going directly to that address you now possess, so there should not be any difference.